nmpc_controller/src/quad_nlp.cpp

```diff
@@ -923,82 +1325,58 @@ void quadNLP::update_solver(
-            }
-          }
-        }
+  for (size_t i = 0; i < N_ - 1; i++) {
+    int idx;
 
-        mu0_ = 1e-1;
-        warm_start_ = false;
-      }
+    if (same_plan_index) {
+      idx = i;
     } else {
-      if (i == N_ - 1) {
+      if (i == N_ - 2) {
         continue;
       }
 
-      if ((contact_sequence_prev.col(i + 1) - contact_sequence_.col(i)).norm() >
-          1e-3) {
-        w0_.block((i + 1) * (n_ + m_) + leg_input_start_idx_, 0,
-                  m_ - leg_input_start_idx_, 1)
-            .fill(0);
-        z_L0_
-            .block((i + 1) * (n_ + m_) + leg_input_start_idx_, 0,
-                   m_ - leg_input_start_idx_, 1)
-            .fill(1);
-        z_U0_
-            .block((i + 1) * (n_ + m_) + leg_input_start_idx_, 0,
-                   m_ - leg_input_start_idx_, 1)
-            .fill(1);
+      idx = i + 1;
+    }
 
-        // Compute the number of contacts
-        double num_contacts = contact_sequence_.col(i).sum();
+    if ((contact_sequence_prev.col(idx) - contact_sequence_.col(i))
+            .cwiseAbs()
+            .sum() > 1e-3) {
+      // Contact change unexpectedly, update the warmstart info
+      get_primal_body_control_var(w0_, idx).fill(0);
+      get_primal_body_control_var(z_L0_, idx).fill(1);
+      get_primal_body_control_var(z_U0_, idx).fill(1);
 
-        // If there are some contacts, set the nominal input accordingly
-        if (num_contacts > 0) {
-          for (size_t j = 0; j < 4; j++) {
-            if (contact_sequence_(j, i) == 1) {
-              w0_((i + 1) * (n_ + m_) + leg_input_start_idx_ + 3 * j + 2, 0) =
-                  mass_ * grav_ / num_contacts;
-            }
-          }
-        }
+      // Compute the number of contactscontact_sequence_
 
-        mu0_ = 1e-1;
-        warm_start_ = false;
-      }
+      mu0_ = 1e-1;
+      warm_start_ = false;
     }
   }
 
-  // Update initial states
-  x_current_ = initial_state.transpose();
+  // If the complexity schedule has changed, update the problem structure
+  bool new_structure =
+      adaptive_complexity_schedule != this->adaptive_complexity_schedule_;
+
+  if (new_structure) {
+    this->adaptive_complexity_schedule_ = adaptive_complexity_schedule;
+    this->update_structure();
+  }
+  // Update initial state
+  x_current_.segment(0, n_body_) = initial_state.head(n_body_);
+  if (n_vec_[0] > n_simple_) {
+    x_current_.segment(n_body_, n_foot_ / 2) = foot_pos_world_.row(0);
+    x_current_.segment(n_body_ + n_foot_ / 2, n_foot_ / 2) =
+        foot_vel_world_.row(0);
+  }
+  if (n_vec_[0] > n_simple_) {
+    x_current_.segment(n_simple_, n_null_) = initial_state.tail(n_null_);
+  }
 
   // Update reference trajectory
   // Local planner has row as N+1 horizon and col as states
-  x_reference_ = ref_traj.transpose();
+  x_reference_.topRows(n_body_) = ref_traj.transpose().topRows(n_body_);
+  x_reference_.middleRows(n_body_, n_foot_ / 2) = foot_pos_world_.transpose();
```

Comment:
Did you add eigen 3.4 to make it support this feature?

Reply:
No this is still in 3.3. Version 3.4 lets you us an array to index into another.

Reply:
I don't think 3.3 has middleRows()

Reply:
It seems to work?